include/boost/url/impl/any_allocator.hpp

```diff
@@ -0,0 +1,190 @@
+#include <boost/config/workaround.hpp>
+#include <boost/type_traits/type_with_alignment.hpp>
+#include <boost/type_traits/make_void.hpp>
+
+#include <cstddef>
+
+namespace boost {
+namespace urls {
+
+namespace detail {
+#if BOOST_WORKAROUND(BOOST_GCC, < 40900)
+    using max_align_t = ::max_align_t;
+#else
+    using max_align_t = std::max_align_t;
+#endif
+
+inline std::size_t ceil_div(std::size_t n, std::size_t d)
+{
+    return (n + d - 1) / d;
+}
+
+template <std::size_t N, class Allocator>
+typename std::enable_if<alignof(max_align_t) <= N, void*>::type
+try_allocate_aligned(const Allocator& a, std::size_t n, std::size_t size, std::size_t align)
+{
+    using aligned_t = max_align_t;
+    using aligned_allocator_t = boost::allocator_rebind_t<Allocator, aligned_t>;
+    return aligned_allocator_t(a).allocate(n * ceil_div(size, align));
+}
+
+template <std::size_t N, class Allocator>
+typename std::enable_if<N < alignof(max_align_t), void*>::type
+try_allocate_aligned(const Allocator& a, std::size_t n, std::size_t size, std::size_t align)
+{
+    if (align <= N)
+    {
+        using aligned_t = typename boost::type_with_alignment<N>::type;
+        using aligned_allocator_t = boost::allocator_rebind_t<Allocator, aligned_t>;
+        return aligned_allocator_t(a).allocate(n * ceil_div(size, align));
+    }
+    else
+    {
+        return try_allocate_aligned<2*N>(a, n, size, align);
+    }
+}
+
+template <class Allocator>
+void* allocate_aligned(const Allocator& a, std::size_t n, std::size_t size, std::size_t align)
+{
+    return try_allocate_aligned<2*alignof(char)>(a, n, size, align);
+}
+
+template <std::size_t N, class Allocator>
+typename std::enable_if<alignof(max_align_t) <= N, void>::type
+try_deallocate_aligned(const Allocator& a, void* p, std::size_t n, std::size_t size, std::size_t align)
+{
+    using aligned_t = max_align_t;
+    using aligned_allocator_t = boost::allocator_rebind_t<Allocator, aligned_t>;
+    aligned_allocator_t(a).deallocate(reinterpret_cast<aligned_t*>(p), n * ceil_div(size, align));
+}
+
+template <std::size_t N, class Allocator>
+typename std::enable_if<N < alignof(max_align_t), void>::type
+try_deallocate_aligned(const Allocator& a, void* p, std::size_t n, std::size_t size, std::size_t align)
+{
+    if (align <= N)
+    {
+        using aligned_t = typename boost::type_with_alignment<N>::type;
+        using aligned_allocator_t = boost::allocator_rebind_t<Allocator, aligned_t>;
+        aligned_allocator_t(a).deallocate(reinterpret_cast<aligned_t*>(p), n * ceil_div(size, align));
+    } else {
+        try_deallocate_aligned<2*N>(a, p, n, size, align);
+    }
+}
+
+template <class Allocator>
+void deallocate_aligned(const Allocator& a, void* p, std::size_t n, std::size_t size, std::size_t align)
+{
+    try_deallocate_aligned<2*alignof(char)>(a, p, n, size, align);
+}
+}
+
+template <class T>
+any_allocator<T>::
+any_allocator() noexcept
+: any_allocator(std::allocator<char>{})
```

Comment:
This can't be noexcept since the _Allocator_ constructor can fail. I don't think allocating memory in the default constructor is a good idea.

Reply:
The required Allocator constructors cannot throw exceptions and the `std::allocator` default constructor is also noexcept.

![image](https://user-images.githubusercontent.com/5369819/149647749-af4d59a0-2897-482d-a90d-9efe9f6cb5f3.png)

Should I still remove it?

Reply:
I don't know

Reply:
It seems ok. If I recall correctly, I think you told me to include the noexcept here. 
Allocator constructors are not expected to throw anyway, std::allocator doesn't throw, and std::allocator is also empty.

Reply:
I said noexcept before Peter suggesting changing the implementation to make p_ always valid. `noexcept` is not correct here, because the 1-arg ctor for `any_allocator` callls `make_shared`, which throws.

Reply:
Mmmm... that's correct. I think making p_ always valid is going to be a problem for us, because including `noexcept` is wrong and removing `noexcept` breaks the named requirements. I suggest we allow `p_` to be `nullptr` again.

Reply:
I agree.

Reply:
Umm... that might not work, lol

Reply:
Conclusion:
it is gcc 4.2 (see above) that requires DefaultConstructible
we should not document any_allocator default ctor, and we should ifdef it just for gcc 4.2 as above
msvc std::string doesn't require default constructible for Allocator, but
msvc std::string allocates memory upon construction in a function marked noexcept

Reply:
The conclusion is this should be default constructible only for GCC < 4.2 and MSVC? Where is that text from?

How this conclusion about the default constructor influences this copy constructor and how does that relate with allowing p_ internally to be nullptr?

Reply:
-> chat